batchflow/models/validator/README.md

```diff
@@ -0,0 +1,224 @@
+- provide API to automized model training and validation.
+
+All you need is:
+
+* define `model_api.py` file with model inherited from `ModelAPI`,
+* configure it by `validator.yaml` file in the same folder,
+* execute the following code from the same folder:
+
+```python
+from batchflow.models.validator import Validator
+
+val = Validator()
+val.start()
+print(val.metrics)
+```
+
+## Basic example
+
+### model_api.py
+
+```python
+import ...
+
+class PascalValidator(ModelAPI):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.model_config = ...
+        self.ds = PascalSegmentation(bar=True)
+
+    def train_loader(self, **kwargs):
+        return self.ds.train
+
+    def test_loader(self, **kwargs):
+        return self.ds.test
+
+    def train(self, ds):
+        train_ppl = ... << ds
+        train_ppl.run()
+        return train_ppl.get_model_by_name('model')
+
+    def inference(self, ds, model):
+        test_ppl = ... << ds
+        test_ppl.run()
+        targets = np.array(test_ppl.v('targets'))
+        predictions = np.array(test_ppl.v('predictions'))
+        return targets, predictions
+```
+
+### validator.yaml
+
+```yaml
+- Pascal Segmentation:
+  - PascalValidator:
+      class: PascalValidator
+      train:
+      test:
+        metrics:
+        - classification:
+            evaluate:
+              - accuracy
+```
+
+### `val.metrics`
+
+```python
+{'Pascal Segmentation':
+       {'PascalValidator':
+              {'classification': {'accuracy': 0.0092724609375}
+       }
+}
+```
+
+## **model_api**
+
+`ModelAPI` child-class can implement the following methods:
+
+### `train_loader(self, path=None, **kwargs)`
+
+`path` and `kwargs` are from config `<task_name>/<model_name>/<train>`. In the example above we have empty value for `train` key, therefore `path=None` and `kwargs={}`. Let's define config:
+
+```yaml
+- Pascal Segmentation:
+  - PascalValidator:
+      class: PascalValidator
+      train:
+        dataset: /path/to/dataset
+```
+
+In that case `path='/path/to/dataset'`. You also can define multiple parameters of `train_loader`:
+
+```yaml
+- Pascal Segmentation:
+  - PascalValidator:
+      class: PascalValidator
+      train:
+         dataset:
+           path: /path/to/dataset
+           format: 'png'
+```
+
+Now `path='/path/to/dataset', kwargs={format: 'png'}`.
+The output of the function will be used as `train_dataset` argument of `train` method. By default, it returns `path`.
+
+### `test_loader(self, path=None, **kwargs)`
+
+The same as `train_loader` but for `test`.
+
+### `load_model(self, path=None, **kwargs)`
+
+Loader for pretrained model. Let's make example above more complex:
+
+```yaml
+- Pascal Segmentation:
+  - PascalValidator:
+      class: PascalValidator
+      train:
+      pretrained:
+        path: /path/to/model
+        device: cuda:0
+      test:
+        metrics:
+        - classification:
+            evaluate:
+              - accuracy
```

Comment:
This is extremely overcomplicated. Make it `metrics: accuracy` on the first nested level (not the sixth). You can lookup for the metrics in the metrics class and if there is a name collision throw an exception. 

`metrics: classification.accuracy` is also possible (also, not instead of)

Reply:
I think that in my `yaml` structure it's easier to understand sense of each parameter because we have two groups of parameters: for `init` and for `evaluate`. 

But I agree that something like `metrics: accuracy` is clearer. I will try to do something like that.

Reply:
@roman-kh what do you think about that?

Example:
```
train:
    dataset: /data/train
test:
    dataset: /data/test
    metrics:
        - f1_score
        - accuracy
f1_score:
    class: classification
    axis: 1
    fmt: 'proba'
    num_classes: 18
    evaluate:
        agg: mean
        multiclass:
```

In `test` we list metrics to compute. If `ModelAPI` class has a method with the same name, it will be computed as a custom metric (`accuracy` here). Otherwise, it will be computed as BatchFlow metric and parameters of that metric we define in additional section (`f1_score`).

Reply:
This look great. However, `f1_score` requires too many parameters. Good defaults are needed here in order to eliminate these parameters for most cases

Reply:
Remove additional nesing for evaluate. 1 level is more than enough for the metrics

Reply:
As for defaults, I think that defaults should be from BatchFlow metric classes. If you can use defaults from BatchFlow, you don't need to configure that metric in yaml.

I think if you want to avoid long lists of parameters in config, you can define those metrics in your class as a custom metric. Config helps to configure exactly bf-metric. That's why I prefer to add an additional level to distinguish `evaluate` and `__init__` methods.